google/cloud/spanner/integration_tests/client_integration_test.cc

```diff
@@ -1076,6 +1078,284 @@ TEST_F(PgClientIntegrationTest, FineGrainedAccessControl) {
+  statements.emplace_back(R"""(
+      CREATE TABLE Customers (
+          CustomerId   INT64 NOT NULL,
+          CustomerName STRING(62) NOT NULL
+      ) PRIMARY KEY (CustomerId))""");
+  statements.emplace_back(R"""(
+      CREATE TABLE ShoppingCarts (
+          CartId       INT64 NOT NULL,
+          CustomerId   INT64 NOT NULL,
+          CustomerName STRING(62) NOT NULL,
+          CONSTRAINT FKShoppingCartsCustomerId
+              FOREIGN KEY (CustomerId)
+              REFERENCES Customers (CustomerId)
+              ON DELETE CASCADE
+      ) PRIMARY KEY (CartId))""");
+  auto create_metadata =
+      admin_client
+          .UpdateDatabaseDdl(GetDatabase().FullName(), std::move(statements))
+          .get();
+  ASSERT_STATUS_OK(create_metadata);
+
+  // ALTER TABLE with ON DELETE CASCADE.
+  statements.clear();
+  statements.emplace_back(R"""(
+      ALTER TABLE ShoppingCarts
+      ADD CONSTRAINT FKShoppingCartsCustomerName
+          FOREIGN KEY (CustomerName)
+          REFERENCES Customers (CustomerName)
+          ON DELETE CASCADE)""");
+  auto add_metadata =
+      admin_client
+          .UpdateDatabaseDdl(GetDatabase().FullName(), std::move(statements))
+          .get();
+  ASSERT_STATUS_OK(add_metadata);
+
+  // Insert a row into the referenced table, and then a referencing row into
+  // the referencing table.
+  auto insert_commit = client_->Commit(
+      Mutations{MakeInsertMutation("Customers", {"CustomerId", "CustomerName"},
+                                   1, "FKCustomer"),
+                MakeInsertMutation("ShoppingCarts",
+                                   {"CartId", "CustomerId", "CustomerName"},  //
+                                   1, 1, "FKCustomer")});
+  ASSERT_STATUS_OK(insert_commit);
+
+  // Cannot insert a row into the referencing table when the key is not
+  // present in the referenced table.
+  auto bad_key = client_->Commit(Mutations{MakeInsertMutation(
+      "ShoppingCarts", {"CartId", "CustomerId", "CustomerName"},  //
+      2, 2, "FKCustomer")});
+  EXPECT_THAT(bad_key, StatusIs(StatusCode::kFailedPrecondition,
+                                AllOf(HasSubstr("Foreign key constraint"),
+                                      HasSubstr("FKShoppingCartsCustomerId"),
+                                      HasSubstr("ShoppingCarts"),
+                                      HasSubstr("Customers(CustomerId)"))));
+
+  // Delete a row in the referenced table. All rows referencing that key
+  // from the referencing table will also be deleted.
+  auto delete_commit = client_->Commit(
+      Mutations{MakeDeleteMutation("Customers", KeySet().AddKey(MakeKey(1)))});
+  ASSERT_STATUS_OK(delete_commit);
+  auto carts =
+      client_->ExecuteQuery(SqlStatement("SELECT CartId FROM ShoppingCarts"));
+  for (auto& cart : StreamOf<std::tuple<std::int64_t>>(carts)) {
+    EXPECT_THAT(cart, IsOk());
+    if (!cart) break;
+    EXPECT_FALSE(true) << "Unexpected cart " << std::get<0>(*cart);
+  }
```

Comment:
comment: This took me a second to understand.

Something like:

```cc
EXPECT_THAT(StreamOf<std::tuple<std::int64_t>>(carts), IsEmpty());
```

... would set a clearer expectation, but the errors in the failure modes would be less useful. Up to you.

Reply:
Unfortunately, it would be more complicated than that, as `StreamOf` returns a "range", which would then have to be used to produce a container that could be matched against `IsEmpty`.

For now then, I choose to leave it as-is, if only to align with the other, similar `StreamOf` expectations.  Perhaps we could later introduce a test helper that converts a stream to a `StatusOr<vector<RowType>>`.